c++/src/kj/compat/url.c++

```diff
@@ -260,21 +250,146 @@ Maybe<Url> Url::tryParse(StringPtr text, Context context, Options options) {
   }
 
   if (text.startsWith("#")) {
-    if (context != REMOTE_HREF) {
-      // No fragment allowed here.
+    if (options.mergeFragment == Options::MergeFragment::NO) {
+      result.fragment = percentDecode(text.slice(1), err, options);
+      text = {};
+    } else {
+      // If we still have a '#', that means that we didn't have a path or a query but still had the fragment.
+      // TODO(now) Should we consider the '#' as part of the authority like we would have for path or query?
       return nullptr;
     }
-    result.fragment = percentDecode(text.slice(1), err, options);
-  } else {
-    // We should have consumed everything.
-    KJ_ASSERT(text.size() == 0);
   }
 
+  // We should have consumed everything.
+  KJ_ASSERT(text.size() == 0, text);
+
   if (err) return nullptr;
 
   return kj::mv(result);
 }
 
+Maybe<Url> Url::tryParse(StringPtr text, Context context, Options options) {
+  switch (context) {
+    case Context::REMOTE_HREF: {
+      if (options.mergeFragment == Options::MergeFragment::AUTO) {
+        options.mergeFragment = Options::MergeFragment::NO;
+      }
+    } break;
+    case Context::HTTP_PROXY_REQUEST: {
+      if (options.mergeFragment == Options::MergeFragment::AUTO) {
+        options.mergeFragment = Options::MergeFragment::YES;
+      }
+    } break;
+    case Context::HTTP_REQUEST: {
+      if (options.mergeFragment == Options::MergeFragment::AUTO) {
+        options.mergeFragment = Options::MergeFragment::YES;
+      }
+    } break;
+    case Context::ARBITRARY: {
+      if (options.mergeFragment == Options::MergeFragment::AUTO) {
+        options.mergeFragment = Options::MergeFragment::NO;
+      }
+    } break;
+  }
+
+  KJ_IF_MAYBE (parsedUrl, tryParseWithoutContext(text, options)) {
+    if (parsedUrl->isValidWithContext(context)) {
+      return kj::mv(*parsedUrl);
+    } else {
+      // TODO(now) Should I leave these logs in? They kinda feel like a poor man's equivalent of
+      // emitting exceptions.
+      KJ_LOG(INFO, "URL failed validation", text, *parsedUrl);
```

Comment:
They weren't here previously AFAICT so I'd vote for removing them.

Reply:
I think I might have been a bit too coy with my TODO. I'm mildly concerned that we provide no channel for explaining why we rejected the URI. I'm guessing this hasn't been an issue in practice but I figured I should check to see if an exception-based form of tryParse would be valuable.